### PR TITLE
feat(model): add Kimi K3 language backbone support

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do?

Adds Megatron Bridge support for the Kimi K3 language backbone, including<br>model/provider construction, HF import, distributed checkpoint save, Megatron<br>greedy inference, HF export mappings, and focused model tests. This remains a<br>draft while strict full-checkpoint HF export and exact round-trip parity are<br>unverified.

# Changelog

* Add the Kimi K3 provider, heterogeneous KDA/MLA layer spec, AttnRes pipeline<br>payload handling, latent-MoE support, and SiTU/KDA operators.
* Register `KimiK3ForConditionalGeneration` / `kimi_k3` with `AutoBridge` and<br>add HF ↔ Megatron parameter mappings.
* Add MXFP4 `uint8`/UE8M0 tensor conversion support.
* Validate and remove the published checkpoint's zero-only KDA `A_log` padding<br>on import, then restore the 32 inactive zero entries on export.
* Preserve the source checkpoint's `vision_tower.*` and `mm_projector.*`<br>tensors unchanged during language-backbone export.
* Extend the toy-HF-model helper to handle nested `text_config`, truncate<br>heterogeneous layer schedules, and download only the safetensor shards<br>required by the selected layers.
* Keep K3 dropout disabled even though the released HF `text_config` does not<br>serialize dropout fields.
* Keep `input_ids` out of the upstream MCore `MoELayer` call. The Miles fork<br>uses this argument for hash routing, while K3 uses sigmoid routing and the<br>MCore version pinned by Bridge does not accept it.
* Add focused unit tests and an agent-readable Kimi K3 model verification card.

# Validation

## Full checkpoint

* Pinned `moonshotai/Kimi-K3` revision<br>`9f62e4e9fffbd0a83ddd60e1c209d828994b3569`.
* GPU HF → Megatron import on 48 GB200 GPUs:
  * TP2/PP3/EP8/ETP2, BF16
  * completed in 27:01
  * persisted `iter_0000000` with 48 distributed-checkpoint shards,<br>`.metadata`, and the 93-layer run configuration
* Megatron greedy inference on 24 GB300 GPUs at clean commit `eb91f8be`:
  * TP1/PP3/EP8/ETP1
  * completed 32 deterministic generation steps in 22:59
  * prompt: `What is 15% of 240? Answer in one short sentence.`
  * exact completion:<br>`15% of 240 = 0.15 × 240 = 36.<|close|>think<|sep|><|open|>response<|sep|>15% of 240 is 36.`
* Source `A_log` audit:
  * 69 tensors, each with 96 active entries and 32 inactive entries
  * all 2,208 inactive values are exactly zero
* Strict full-checkpoint export is not claimed:
  * the checkpoint contains 497,220 source tensors
  * TP4/PP3/EP8/ETP4 reduced each rank to about 34–35B parameters and<br>eliminated all GPU-to-CPU merge fallbacks
  * the 96-GB200 run had no traceback, OOM, or NCCL error, but the full export<br>generator did not complete within the one-hour allocation
  * strict HF reload and exact tensor audit therefore remain pending

## Focused tests

* `uv run --no-project --with pre-commit pre-commit run --all-files`: passed.
* Focused Kimi/quantization tests: 30 passed.
* Export-padding and multimodal passthrough tests: 9 passed.
* Four-layer proxy generated from the pinned checkpoint:
  * layers `0..3` of `93`
  * 7 source shards, 16,402 tensors, 58,480,464,384 tensor bytes
* TP8/EP8 import and train-step smoke on 8 H100 GPUs:
  * model construction and HF import tasks completed
  * real LM loss forward/backward and SGD update completed
  * loss `13.93112946`; 769 parameters received gradients
* TP8/EP8 RP2 learnability smokes on 8 H100 GPUs:
  * fixed-window overfit: `13.92106724` → `0.00057347` in 6 steps
  * 100-window rolling run: first-10 mean `13.56379385`, last-10 mean<br>`12.54698963` (`-7.4965%`)
* TP8/EP8/PP2 state validation on 16 H100 GPUs completed model construction,<br>import, collectives, and `sharded_state_dict()`.

# Known draft gaps

* Strict full-checkpoint Megatron → HF export, HF reload, and exact round-trip<br>parity remain unverified.
* Full HF/Megatron forward-logit correlation remains unverified; the completed<br>four-layer proxy reached cosine similarity `0.9998329878` and Pearson<br>correlation `0.9998273849`.
* Production pretraining/SFT/PEFT recipes, checkpoint-resume validation,<br>performance tuning, and user documentation remain pending.
* VPP is not supported yet, and KDA with CP greater than 1 remains unsupported.
* This phase covers the language backbone. Native K3 vision/video modeling and<br>multimodal inference are not included; export only preserves the published<br>vision/projector tensors unchanged.

# Original implementation references

The initial K3 modeling and mapping implementation was adapted from the<br>Apache-2.0 Miles Kimi K3 backend:

* [Miles Kimi K3 PR #1825](<https://github.com/radixark/miles/pull/1825>)
* [Original Megatron Bridge mapping at commit `dc62a0b`](<https://github.com/radixark/miles/blob/dc62a0bd4b7af1c59ee2084852eb18b5585ec082/miles_plugins/mbridge/kimi_k3.py>)
* [Original K3 modeling files at commit `dc62a0b`](<https://github.com/radixark/miles/tree/dc62a0bd4b7af1c59ee2084852eb18b5585ec082/miles_plugins/models/kimi_k3>)

# Before this PR is ready for review

- [X] Read and followed the contributor guidelines.
- [X] Added focused unit tests.
- [X] Added a model verification card with verified import and inference.
- [ ] Complete strict full-checkpoint export and exact round-trip parity.
- [ ] Add or update the remaining user documentation.
- [X] Optional-component import behavior is unchanged by this draft.

# Additional Information

* Related to NVIDIA-NeMo/Megatron-Bridge#4910.

### CODE REVIEW

**Comment by yaoyu-33** [comment]
Full-model MCore greedy inference validation on Lyris:

- Hardware/parallelism: 8 GB300 nodes, 24 ranks, TP1 / PP3 / EP8 / ETP1.
- Before this fix, the first forward failed at the K3 AttnRes pipeline boundary: `stage-boundary payload width 7168 != (1 + 3) * 7168`.
- Enabling `variable_seq_lengths` makes MCore exchange the actual packed activation shape at each PP boundary (4x hidden at the first boundary and 7x hidden at the second).
- PP2/EP2 4-layer proxy completed one greedy step successfully (Slurm 2512806, exit 0).
- Full Kimi-K3 completed all 32 greedy argmax steps (Slurm 2512803, exit 0). Prompt: `What is 15% of 240? Answer in one short sentence.` Output: `15% of 240 = 0.15 × 240 = 36.<|close|>think<|sep|><|open|>response<|sep|>15% of 240 is 36.`
- Focused unit tests: 5 passed.
- `pre-commit run --all-files`: passed.

**Comment by yaoyu-33** [comment]
Added the Kimi K3 model verification card in commit `d4cbc9a38` at `examples/model_verification_cards/kimi-k3/card.yaml`.

The card follows the strict verification gates: the four persisted conversion paths, full-model forward parity, inference clean-commit rerun, and all training workflows remain unverified. It records the completed full-model direct-import/greedy-generation result as diagnostic evidence only, and documents the KDA `A_log` 96-active-head versus 128-stored-entry export gap. Training is intentionally deferred for this model-only pass.

Validation completed:
- model-card validator passed
- `git diff --check` passed
- full pre-commit suite passed

File: examples/conversion/create_hf_toy_model.py [hunk 1]
```diff
@@ -127,27 +189,49 @@ def _copy_metadata_files(source_dir: Path, output_dir: Path) -> None:
...
+            layer for layer in transformer_config["mlp_only_layers"] if layer < num_hidden_layers
+        ]
+    linear_attn_config = transformer_config.get("linear_attn_config")
+    if isinstance(linear_attn_config, dict):
+        for layer_list_name in ("kda_layers", "full_attn_layers"):
+            if isinstance(linear_attn_config.get(layer_list_name), list):
+                linear_attn_config[layer_list_name] = [
+                    layer_number
+                    for layer_number in linear_attn_config[layer_list_name]
+                    if layer_number <= num_hidden_layers
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply examples/conversion/create_hf_toy_model.py:240]: Confirmed 1-indexed, so the inclusive bound is correct — and added a comment saying so in `31f8fa102`.

Checked against the pinned revision `9f62e4e9`: `num_hidden_layers` is 93, `kda_layers` has 69 entries spanning 1..91, `full_attn_layers` has 24 entries spanning 4..93. 69 + 24 = 93 and the two partition 1..93 exactly, so there is no layer 0 and the maximum is `num_hidden_layers` itself. The consumer side agrees — `KimiK3Attention.__init__` tests `layer_number in config.kimi_kda_layers` against MCore's 1-indexed `layer_number`.

**Comment by yaoyu-33** [comment]
Thanks — addressed in `d708406c3` and `31f8fa102`.

**1. `linear_attn_config` layer-list bound** — confirmed 1-indexed, `<=` is correct; added the clarifying comment. Details in the [inline reply](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5130#discussion_r3676920257).

**2. `flop_utils.py` / KDA** — fixed rather than deferred. `KimiK3ModelProvider` subclasses `MLAModelProvider`, so `num_floating_point_operations` was costing all 93 layers as full MLA including the quadratic core-attention term, when 69 of them are KDA linear-attention blocks. Added a KDA branch that re-costs the layers in `kimi_kda_layers` with the KDA per-layer formula (q/k/v/gate/output projections, per-head beta, low-rank forget gate, depthwise short convs over q/k/v, and the chunked delta-rule recurrence over the `[head_dim, head_dim]` state), mirroring the existing `gated_delta_net` override. MTP layers inherit the final decoder layer's attention type, and out-of-range layer numbers raise.

Eight new tests in `TestKimiK3KdaFlops` cover the exact weighted self-attn term, MTP type inheritance, the out-of-range guard, and that an all-KDA schedule scales exactly linearly in sequence length. `tests/unit_tests/training/utils/test_flop_utils.py` + `tests/unit_tests/models/kimi/` + `tests/unit_tests/models/test_quantization_utils.py`: **153 passed** (nemo:26.06.01-based container).

**3. Docs / examples / recipe gap** — added `docs/models/kimi/kimi-k3.md` (supported variants, the heterogeneous KDA/MLA schedule, MXFP4 dequant, `A_log` padding round-trip, conversion command, and known limitations), registered in `docs/models/kimi/index.md` and `docs/models/README.md`. The page carries an explicit in-progress note pointing at the verification card.

The `examples/models/kimi/k3/` scripts and the training recipe are **intentionally deferred**, as the PR body's Known draft gaps state: the skill's acceptance bar for `conversion.sh` / `inference.sh` is a clean round-trip on a real published model, and strict full-checkpoint export plus exact round-trip parity are still unverified here. Shipping a recipe now would also mean publishing parallelism settings that no training run has validated. Both land once export parity is closed.
